source/extensions/filters/http/cache/thundering_herd_handler.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include "envoy/extensions/filters/http/cache/v3/cache.pb.h"
+#include "envoy/http/filter.h"
+
+#include "source/extensions/filters/http/cache/http_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+class CacheFilter;
+
+class ThunderingHerdRetryInterface {
+public:
+  virtual void retryHeaders(Http::RequestHeaderMap& request_headers) PURE;
+  virtual ~ThunderingHerdRetryInterface() = default;
+};
+
+class ThunderingHerdHandler {
+public:
+  static std::shared_ptr<ThunderingHerdHandler> create(
+      const envoy::extensions::filters::http::cache::v3::CacheConfig::ThunderingHerdHandler& config,
+      Server::Configuration::CommonFactoryContext& context);
+
+  /**
+   * Either continues encoding, or blocks waiting for a previous request to complete
+   * before retrying the cache lookup.
+   * @param weak_filter the filter to be blocked or continued.
+   * @param decoder_callbacks the decoder_callbacks from the filter.
+   * @param key the key of the cache entry.
+   * @param request_headers the request headers from the filter.
+   */
+  virtual void handleUpstreamRequest(std::weak_ptr<ThunderingHerdRetryInterface> weak_filter,
+                                     Http::StreamDecoderFilterCallbacks* decoder_callbacks,
```

Comment:
CacheFilter shouldn't pass its decoder_callbacks to ThunderingHerdHandler, because CacheFilter is responsible for ensuring that nothing calls a callback after CacheFilter::OnDestroy is called. Instead, CacheFilter could expose an "unblockDecoding" method that includes something like:
ASSERT(thundering_herd_in_flight_);
if(filter_state_ == destroyed) return;